ql/lib/codeql/kaleidoscope/Cfg.qll

```diff
@@ -0,0 +1,254 @@
+private import codeql.kaleidoscope.controlflow.BasicBlocks
+
+private module Completion {
+  private newtype TCompletion =
+    TSimpleCompletion() or
+    TBooleanCompletion(boolean b) { b in [false, true] } or
+    TReturnCompletion()
+
+  abstract class Completion extends TCompletion {
+    abstract string toString();
+
+    predicate isValidForSpecific(AstNode e) { none() }
+
+    predicate isValidFor(AstNode e) { this.isValidForSpecific(e) }
+
+    abstract SuccessorType getAMatchingSuccessorType();
+  }
+
+  class SimpleCompletion extends Completion, TSimpleCompletion {
+    override string toString() { result = "SimpleCompletion" }
+
+    override predicate isValidFor(AstNode e) { not any(Completion c).isValidForSpecific(e) }
+
+    override NormalSuccessor getAMatchingSuccessorType() { any() }
+  }
+
+  class BooleanCompletion extends Completion, TBooleanCompletion {
+    boolean value;
+
+    BooleanCompletion() { this = TBooleanCompletion(value) }
+
+    override string toString() { result = "BooleanCompletion(" + value + ")" }
+
+    override predicate isValidForSpecific(AstNode e) {
+      e = any(ConditionalExpression c).getCondition() or
+      e = any(ForExpression c).getCondition()
+    }
+
+    override BooleanSuccessor getAMatchingSuccessorType() { result.getValue() = value }
+
+    final boolean getValue() { result = value }
+  }
+
+  class ReturnCompletion extends Completion, TReturnCompletion {
+    override string toString() { result = "ReturnCompletion" }
+
+    override predicate isValidForSpecific(AstNode e) { none() }
+
+    override ReturnSuccessor getAMatchingSuccessorType() { any() }
+  }
+
+  cached
+  private newtype TSuccessorType =
+    TNormalSuccessor() or
+    TBooleanSuccessor(boolean b) { b in [false, true] } or
+    TReturnSuccessor()
+
+  class SuccessorType extends TSuccessorType {
+    string toString() { none() }
+  }
+
+  class NormalSuccessor extends SuccessorType, TNormalSuccessor {
+    override string toString() { result = "successor" }
+  }
+
+  class BooleanSuccessor extends SuccessorType, TBooleanSuccessor {
+    boolean value;
+
+    BooleanSuccessor() { this = TBooleanSuccessor(value) }
+
+    override string toString() { result = value.toString() }
+
+    boolean getValue() { result = value }
+  }
+
+  class ReturnSuccessor extends SuccessorType, TReturnSuccessor {
+    override string toString() { result = "return" }
+  }
+}
+
+private module CfgScope {
+  abstract class CfgScope extends AstNode { }
+
+  private class FunctionScope extends CfgScope, FunctionDefinition { }
+
+  private class ProgramScope extends CfgScope, Program { }
+}
+
+private module Implementation implements CfgShared::InputSig<Location> {
+  import codeql.kaleidoscope.Ast
+  import Completion
+  import CfgScope
+
+  predicate completionIsNormal(Completion c) { not c instanceof ReturnCompletion }
+
+  // Not using CFG splitting, so the following are just a dummy types.
+  private newtype TUnit = Unit()
```

Comment:
May as well use the one from the `util` pack.

Reply:
That's true, but it felt it bit much to import another library for just this line.